File: lab 3/FileManager.cpp
#include "FileManager.h"
#include "SingleEvent.h"
#include "RecurringEvent.h"
#include <fstream>
#include <iostream>
using namespace std;

bool FileManager::saveEventsToFile(const vector<Event*>& events, const string& filename)
{
    ofstream outFile(filename);
    if (!outFile) return false;

    for (const auto& event : events)
    {
        auto singleEvent = dynamic_cast<SingleEvent*>(event);
        if (singleEvent) {
            outFile << "SingleEvent "
                << singleEvent->getName() << " "
                << singleEvent->getDuration() << " "
                << singleEvent->getDate()[0] << " "
                << singleEvent->getDate()[1] << " "
                << singleEvent->getDate()[2] << endl;
        }
        else {
            auto recurringEvent = dynamic_cast<RecurringEvent*>(event);
            if (recurringEvent) {
                outFile << "RecurringEvent "
                    << recurringEvent->getName() << " "
                    << recurringEvent->getDuration() << " "
                    << recurringEvent->getStartDate()[0] << " "
                    << recurringEvent->getStartDate()[1] << " "
                    << recurringEvent->getStartDate()[2] << " "
                    << recurringEvent->getEndDate()[0] << " "
                    << recurringEvent->getEndDate()[1] << " "
                    << recurringEvent->getEndDate()[2] << " "
                    << recurringEvent->getRepeats() << endl;
            }
        }
    }

    return true;
}


bool FileManager::loadEventsFromFile(vector<Event*>& events, const string& filename)
{
    ifstream inFile(filename);
    if (!inFile) return false;

    events.clear();
    string type;

    while (inFile >> type) {
        if (type == "SingleEvent") {
            string name;
            double duration;
            int day, month, year;

            inFile >> ws; 
            getline(inFile, name, ' ');
            inFile >> duration >> day >> month >> year;

            events.push_back(new SingleEvent(name, duration, day, month, year));
        }
        else if (type == "RecurringEvent") {
            string name;
            double duration;
            int sDay, sMonth, sYear, eDay, eMonth, eYear, repeats;

            inFile >> ws; 
            getline(inFile, name, ' ');
            inFile >> duration >> sDay >> sMonth >> sYear
                >> eDay >> eMonth >> eYear >> repeats;

            events.push_back(new RecurringEvent(name, duration, sDay, sMonth, sYear, eDay, eMonth, eYear, repeats));
        }
    }

    return true;
}

File: python/FileManager.py
from SingleEvent import SingleEvent
from ReccuringEvent import ReccuringEvent

class FileManager:


    def save_events_to_file(events, filename):
        with open(filename, 'w') as out_file:
            for event in events:
                if isinstance(event, SingleEvent):
                    out_file.write(f"SingleEvent {event.get_name()} {event.get_duration()} {event.get_date()[0]} {event.get_date()[1]} {event.get_date()[2]}\n")
                elif isinstance(event, ReccuringEvent):
                    out_file.write(f"ReccuringEvent {event.get_name()} {event.get_duration()} {event.get_start_date()[0]} {event.get_start_date()[1]} {event.get_start_date()[2]} "
                                   f"{event.get_end_date()[0]} {event.get_end_date()[1]} {event.get_end_date()[2]} {event.get_repeats()}\n")
        return True


    def load_events_from_file(events, filename):
        with open(filename, 'r') as in_file:
            events.clear()
            for line in in_file:
                parts = line.strip().split()
                event_type = parts[0]

                if event_type == "SingleEvent":
                    name = parts[1]
                    duration = float(parts[2])
                    day, month, year = map(int, parts[3:6])
                    events.append(SingleEvent().change_event(name, duration, day, month, year))

                elif event_type == "ReccuringEvent":
                    name = parts[1]
                    duration = float(parts[2])
                    s_day, s_month, s_year = map(int, parts[3:6])
                    e_day, e_month, e_year = map(int, parts[6:9])
                    repeats = int(parts[9])
                    events.append(ReccuringEvent().change_event(name, duration, s_day, s_month, s_year, e_day, e_month, e_year, repeats))
        return True
